api/pubsub.py

```diff
@@ -100,9 +110,12 @@ async def subscribe(self, channel, user):
         async with self._lock:
             redis_sub = self._redis.pubsub()
             sub = Subscription(id=sub_id, channel=channel, user=user)
-            self._subscriptions[sub_id] = {'redis_sub': redis_sub,
-                                           'sub': sub}
             await redis_sub.subscribe(channel)
+            await self._redis.client_setname(user)
```

Comment:
`client_setname` will be executed in the context of a current connection. So, it won't set a global name for all the connections. So, the `_redis` denotes the current connection. In addition to its [documentation](https://redis.io/commands/client-setname/) I checked some of the GH issues and PR of the package.
Please see https://github.com/redis/redis-py/issues/2245#issuecomment-1163252001 and https://github.com/redis/redis-py/pull/2247/files

I think this also explains why l am getting `idle` time of the last subscriber always `0`. It's because `_redis.client_list` command uses the last connection instance to get client information. Hence, the last executed command is also `client` for the connection.

Reply:
Yes, I am aware of the exact problems you have described here and already trying to dig deeper into the Redis source code and documentation to get an understanding.
Thanks for describing issues in much detail though.

Reply:
So, I investigated and checked different client IDs after each operation and this is what I got:

```
kernelci-api | Client ID after creating redis client with `aioredis.from_url`: 3
kernelci-api | INFO:     172.20.0.1:34926 - "GET /latest/users HTTP/1.1" 401 Unauthorized

kernelci-api | In PubSub.subscriber function
kernelci-api | Current client ID: 3
kernelci-api | Cliend ID after redis_sub.subscribe :4
kernelci-api | client_info <redis.asyncio.client.PubSub object at 0x7f09fd1172b0> {'id': 4, 'addr': '172.20.0.6:56510', 'laddr': '172.20.0.3:6379', 'fd': '9', 'name': 'admin', 'age': 0, 'idle': 0, 'flags': 'N', 'db': 1, 'sub': 0, 'psub': 0, 'multi': -1, 'qbuf': 26, 'qbuf-free': 40928, 'argv-mem': 10, 'obl': 0, 'oll': 0, 'omem': 0, 'tot-mem': 61466, 'events': 'r', 'cmd': 'client', 'user': 'default', 'redir': '-1'}
kernelci-api | self._subscriptions--- {754: {'redis_sub': <redis.asyncio.client.PubSub object at 0x7f09fd1172b0>, 'sub': Subscription(id=754, channel='user-admin1', user='admin'), 'client_id': 4}}
kernelci-api | INFO:     172.20.0.1:42734 - "POST /latest/subscribe/user-admin1 HTTP/1.1" 200 OK


kernelci-api | In PubSub.subscription_stats function
kernelci-api | Current client ID: 4
kernelci-api | get all pubsub channels: [b'user-admin1']
kernelci-api | all clients: [{'id': '3', 'addr': '172.20.0.6:46586', 'laddr': '172.20.0.3:6379', 'fd': '8', 'name': '', 'age': '32', 'idle': '18', 'flags': 'P', 'db': '1', 'sub': '1', 'psub': '0', 'multi': '-1', 'qbuf': '0', 'qbuf-free': '0', 'argv-mem': '0', 'obl': '0', 'oll': '0', 'omem': '0', 'tot-mem': '20504', 'events': 'r', 'cmd': 'subscribe', 'user': 'default', 'redir': '-1'}, {'id': '4', 'addr': '172.20.0.6:56510', 'laddr': '172.20.0.3:6379', 'fd': '9', 'name': 'admin', 'age': '18', 'idle': '0', 'flags': 'N', 'db': '1', 'sub': '0', 'psub': '0', 'multi': '-1', 'qbuf': '26', 'qbuf-free': '40928', 'argv-mem': '10', 'obl': '0', 'oll': '0', 'omem': '0', 'tot-mem': '61466', 'events': 'r', 'cmd': 'client', 'user': 'default', 'redir': '-1'}]
kernelci-api | pubsub clients:  [{'id': '3', 'addr': '172.20.0.6:46586', 'laddr': '172.20.0.3:6379', 'fd': '8', 'name': '', 'age': '32', 'idle': '18', 'flags': 'P', 'db': '1', 'sub': '1', 'psub': '0', 'multi': '-1', 'qbuf': '0', 'qbuf-free': '0', 'argv-mem': '0', 'obl': '0', 'oll': '0', 'omem': '0', 'tot-mem': '20504', 'events': 'r', 'cmd': 'subscribe', 'user': 'default', 'redir': '-1'}]
kernelci-api | normal clients:  [{'id': '4', 'addr': '172.20.0.6:56510', 'laddr': '172.20.0.3:6379', 'fd': '9', 'name': 'admin', 'age': '18', 'idle': '0', 'flags': 'N', 'db': '1', 'sub': '0', 'psub': '0', 'multi': '-1', 'qbuf': '48', 'qbuf-free': '40906', 'argv-mem': '20', 'obl': '0', 'oll': '0', 'omem': '0', 'tot-mem': '61492', 'events': 'r', 'cmd': 'client', 'user': 'default', 'redir': '-1'}]
kernelci-api | self._subscriptions--- {754: {'redis_sub': <redis.asyncio.client.PubSub object at 0x7f09fd1172b0>, 'sub': Subscription(id=754, channel='user-admin1', user='admin'), 'client_id': 4}}
kernelci-api | INFO:     172.20.0.1:51662 - "GET /latest/stats/subscriptions HTTP/1.1" 200 OK


kernelci-api | In PubSub.unsubscribe
kernelci-api | Current client ID: 4
kernelci-api | unsubscribing--- <redis.asyncio.client.PubSub object at 0x7f09fd1172b0>
kernelci-api | INFO:     172.20.0.1:60758 - "POST /latest/unsubscribe/754 HTTP/1.1" 200 OK


kernelci-api | In PubSub.subscription_stats
kernelci-a:pi | Current client ID: 4
kernelci-api | all channels []
kernelci-api | all clients: [{'id': '3', 'addr': '172.20.0.6:46586', 'laddr': '172.20.0.3:6379', 'fd': '8', 'name': '', 'age': '162', 'idle': '13', 'flags': 'N', 'db': '1', 'sub': '0', 'psub': '0', 'multi': '-1', 'qbuf': '0', 'qbuf-free': '0', 'argv-mem': '0', 'obl': '0', 'oll': '0', 'omem': '0', 'tot-mem': '20496', 'events': 'r', 'cmd': 'unsubscribe', 'user': 'default', 'redir': '-1'}, {'id': '4', 'addr': '172.20.0.6:56510', 'laddr': '172.20.0.3:6379', 'fd': '9', 'name': 'admin', 'age': '148', 'idle': '0', 'flags': 'N', 'db': '1', 'sub': '0', 'psub': '0', 'multi': '-1', 'qbuf': '26', 'qbuf-free': '40928', 'argv-mem': '10', 'obl': '0', 'oll': '0', 'omem': '0', 'tot-mem': '61466', 'events': 'r', 'cmd': 'client', 'user': 'default', 'redir': '-1'}]
kernelci-api | pubsub clients:  []
kernelci-api | normal clients:   [{'id': '3', 'addr': '172.20.0.6:46586', 'laddr': '172.20.0.3:6379', 'fd': '8', 'name': '', 'age': '162', 'idle': '13', 'flags': 'N', 'db': '1', 'sub': '0', 'psub': '0', 'multi': '-1', 'qbuf': '0', 'qbuf-free': '0', 'argv-mem': '0', 'obl': '0', 'oll': '0', 'omem': '0', 'tot-mem': '20496', 'events': 'r', 'cmd': 'unsubscribe', 'user': 'default', 'redir': '-1'}, {'id': '4', 'addr': '172.20.0.6:56510', 'laddr': '172.20.0.3:6379', 'fd': '9', 'name': 'admin', 'age': '148', 'idle': '0', 'flags': 'N', 'db': '1', 'sub': '0', 'psub': '0', 'multi': '-1', 'qbuf': '48', 'qbuf-free': '40906', 'argv-mem': '20', 'obl': '0', 'oll': '0', 'omem': '0', 'tot-mem': '61492', 'events': 'r', 'cmd': 'client', 'user': 'default', 'redir': '-1'}]
kernelci-api | self._subscriptions--- {}
```

Summary:

Client ID of Redis client connection to server using `aioredis.from_url` is `3`. 
After subscribing to a channel, a new client is created with client ID `4`.
Based on the client IDs after each operation, it seems like it always uses the last connected client ID to execute redis commands. For instance, it uses client ID `4` to get the list of clients `redis.client_list` in `subscription_stats` function. Even after unsubscribing, it still takes its client ID (i.e. 4) in stats endpoint. And that's weird.
I looked into `redis-py` source code and tests implementation to know its internals. However, I didn't get any clear idea from it. The documentation also doesn't specify anything about it. So questions remained unanswered here.

Reply:
Based on the above summary, it is not possible to get the correct `idle` time for a pubsub client as it will change based on the last Redis command instead of the actual poll time. The only option now is to handle it manually. I'll update the PR.